server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java

```diff
@@ -1846,4 +1846,25 @@ public void test_generated_sub_column_provided_null_is_validated() {
         execute("insert into t (id, obj) (select 1, {\"a\" = {\"b\" = null}} from sys.cluster)");
         assertThat(response.rowCount()).isEqualTo(0L);
     }
+
+    @Test
+    public void test_generated_column_can_refer_to_a_non_string_partitioned_by_column() {
+        execute("""
+            create table t(a int, parted int, gen_from_parted int as parted +1) partitioned by (parted);
+            """
+        );
+
+        // insert from values
+        execute("insert into t (a, parted) values (1, 2)");
+        refresh();
+        execute("select a, parted, gen_from_parted from t");
+        assertThat(response).hasRows("1| 2| 3");
+
+        // insert from sub-query
+        execute("insert into t (a, parted) select 2, 3");
+        assertThat(response.rowCount()).isEqualTo(1L);
+        refresh();
+        execute("select a, parted, gen_from_parted from t order by a");
+        assertThat(response).hasRows("1| 2| 3", "2| 3| 4");
+    }
```

Comment:
Checked that scenario on 5.2.8 and also got 0 rows in insert from subquery. 

It's not a regression on 5.3.0. 

Debugged on 5.2 branch and apparently record rejected with ClassCastException in `InsertSourceFromCells`, somewhere in `IntegerType.sanitizeValue`.